<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 7 多组学分析 | R cookbook from Xiaotao Shen</title>
  <meta name="description" content="记录所有跟R语言有关的内容" />
  <meta name="generator" content="bookdown 0.18 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 7 多组学分析 | R cookbook from Xiaotao Shen" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="记录所有跟R语言有关的内容" />
  <meta name="github-repo" content="jaspershen/r_cookbook" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 7 多组学分析 | R cookbook from Xiaotao Shen" />
  
  <meta name="twitter:description" content="记录所有跟R语言有关的内容" />
  

<meta name="author" content="  Dr. Xiaotao Shen   Postdoctoral at Snyder lab   Stanford University School of Medicine      shenxt.info    shenxt@stanford.edu" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="network.html"/>
<link rel="next" href="regularexpression.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">R cookbook of Xiaotao Shen</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="chapter" data-level="1" data-path="markdown.html"><a href="markdown.html"><i class="fa fa-check"></i><b>1</b> Markdown</a><ul>
<li class="chapter" data-level="1.1" data-path="markdown.html"><a href="markdown.html#markdown语法"><i class="fa fa-check"></i><b>1.1</b> Markdown语法</a></li>
<li class="chapter" data-level="1.2" data-path="markdown.html"><a href="markdown.html#标题-header"><i class="fa fa-check"></i><b>1.2</b> 标题 (Header)</a></li>
<li class="chapter" data-level="1.3" data-path="markdown.html"><a href="markdown.html#图片-image"><i class="fa fa-check"></i><b>1.3</b> 图片 (Image)</a></li>
<li class="chapter" data-level="1.4" data-path="markdown.html"><a href="markdown.html#链接-links"><i class="fa fa-check"></i><b>1.4</b> 链接 (Links)</a></li>
<li class="chapter" data-level="1.5" data-path="markdown.html"><a href="markdown.html#强调-emphasize"><i class="fa fa-check"></i><b>1.5</b> 强调 (Emphasize)</a></li>
<li class="chapter" data-level="1.6" data-path="markdown.html"><a href="markdown.html#引用blockquotes"><i class="fa fa-check"></i><b>1.6</b> 引用(Blockquotes)</a></li>
<li class="chapter" data-level="1.7" data-path="markdown.html"><a href="markdown.html#代码块-code-chunks"><i class="fa fa-check"></i><b>1.7</b> 代码块 (Code chunks)</a></li>
<li class="chapter" data-level="1.8" data-path="markdown.html"><a href="markdown.html#水平线-horizontal-rule"><i class="fa fa-check"></i><b>1.8</b> 水平线 (Horizontal rule)</a></li>
<li class="chapter" data-level="1.9" data-path="markdown.html"><a href="markdown.html#列表-list"><i class="fa fa-check"></i><b>1.9</b> 列表 (List)</a></li>
<li class="chapter" data-level="1.10" data-path="markdown.html"><a href="markdown.html#表格-tables"><i class="fa fa-check"></i><b>1.10</b> 表格 (Tables)</a></li>
<li class="chapter" data-level="1.11" data-path="markdown.html"><a href="markdown.html#段落"><i class="fa fa-check"></i><b>1.11</b> 段落</a></li>
<li class="chapter" data-level="1.12" data-path="markdown.html"><a href="markdown.html#插入icon"><i class="fa fa-check"></i><b>1.12</b> 插入icon</a></li>
<li class="chapter" data-level="1.13" data-path="markdown.html"><a href="markdown.html#数学公式-equations"><i class="fa fa-check"></i><b>1.13</b> 数学公式 (Equations)</a><ul>
<li class="chapter" data-level="1.13.1" data-path="markdown.html"><a href="markdown.html#行内和行外"><i class="fa fa-check"></i><b>1.13.1</b> 行内和行外</a></li>
<li class="chapter" data-level="1.13.2" data-path="markdown.html"><a href="markdown.html#上标和下标"><i class="fa fa-check"></i><b>1.13.2</b> 上标和下标</a></li>
<li class="chapter" data-level="1.13.3" data-path="markdown.html"><a href="markdown.html#占位符"><i class="fa fa-check"></i><b>1.13.3</b> 占位符</a></li>
<li class="chapter" data-level="1.13.4" data-path="markdown.html"><a href="markdown.html#四则运算"><i class="fa fa-check"></i><b>1.13.4</b> 四则运算</a></li>
<li class="chapter" data-level="1.13.5" data-path="markdown.html"><a href="markdown.html#集合运算"><i class="fa fa-check"></i><b>1.13.5</b> 集合运算</a></li>
<li class="chapter" data-level="1.13.6" data-path="markdown.html"><a href="markdown.html#数学符号"><i class="fa fa-check"></i><b>1.13.6</b> 数学符号</a></li>
<li class="chapter" data-level="1.13.7" data-path="markdown.html"><a href="markdown.html#希腊字母"><i class="fa fa-check"></i><b>1.13.7</b> 希腊字母</a></li>
</ul></li>
<li class="chapter" data-level="1.14" data-path="markdown.html"><a href="markdown.html#视频-video和音乐-music"><i class="fa fa-check"></i><b>1.14</b> 视频 (Video)和音乐 (music)</a></li>
<li class="chapter" data-level="1.15" data-path="markdown.html"><a href="markdown.html#markdown编辑器"><i class="fa fa-check"></i><b>1.15</b> Markdown编辑器</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="rmarkdown.html"><a href="rmarkdown.html"><i class="fa fa-check"></i><b>2</b> Rmarkdown</a><ul>
<li class="chapter" data-level="2.1" data-path="rmarkdown.html"><a href="rmarkdown.html#安装"><i class="fa fa-check"></i><b>2.1</b> 安装</a></li>
<li class="chapter" data-level="2.2" data-path="rmarkdown.html"><a href="rmarkdown.html#rmarkdown结构"><i class="fa fa-check"></i><b>2.2</b> Rmarkdown结构</a></li>
<li class="chapter" data-level="2.3" data-path="rmarkdown.html"><a href="rmarkdown.html#yaml"><i class="fa fa-check"></i><b>2.3</b> YAML</a></li>
<li class="chapter" data-level="2.4" data-path="rmarkdown.html"><a href="rmarkdown.html#r代码块"><i class="fa fa-check"></i><b>2.4</b> R代码块</a></li>
<li class="chapter" data-level="2.5" data-path="rmarkdown.html"><a href="rmarkdown.html#代码块参数-chunk-options"><i class="fa fa-check"></i><b>2.5</b> 代码块参数 (Chunk options)</a><ul>
<li class="chapter" data-level="2.5.1" data-path="rmarkdown.html"><a href="rmarkdown.html#代码是否执行"><i class="fa fa-check"></i><b>2.5.1</b> 代码是否执行</a></li>
<li class="chapter" data-level="2.5.2" data-path="rmarkdown.html"><a href="rmarkdown.html#代码控制-code-decoration"><i class="fa fa-check"></i><b>2.5.2</b> 代码控制 (code decoration)</a></li>
<li class="chapter" data-level="2.5.3" data-path="rmarkdown.html"><a href="rmarkdown.html#缓冲-cache"><i class="fa fa-check"></i><b>2.5.3</b> 缓冲 (cache)</a></li>
<li class="chapter" data-level="2.5.4" data-path="rmarkdown.html"><a href="rmarkdown.html#图像控制plot"><i class="fa fa-check"></i><b>2.5.4</b> 图像控制(plot)</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="rmarkdown.html"><a href="rmarkdown.html#全局代码块参数"><i class="fa fa-check"></i><b>2.6</b> 全局代码块参数</a></li>
<li class="chapter" data-level="2.7" data-path="rmarkdown.html"><a href="rmarkdown.html#具体问题的cookbook"><i class="fa fa-check"></i><b>2.7</b> 具体问题的cookbook</a></li>
<li class="chapter" data-level="2.8" data-path="rmarkdown.html"><a href="rmarkdown.html#rmarkdown特有语法"><i class="fa fa-check"></i><b>2.8</b> Rmarkdown特有语法</a><ul>
<li class="chapter" data-level="2.8.1" data-path="rmarkdown.html"><a href="rmarkdown.html#图片-image-1"><i class="fa fa-check"></i><b>2.8.1</b> 图片 (image)</a></li>
<li class="chapter" data-level="2.8.2" data-path="rmarkdown.html"><a href="rmarkdown.html#表格-table"><i class="fa fa-check"></i><b>2.8.2</b> 表格 (Table)</a></li>
</ul></li>
<li class="chapter" data-level="2.9" data-path="rmarkdown.html"><a href="rmarkdown.html#输出格式"><i class="fa fa-check"></i><b>2.9</b> 输出格式</a><ul>
<li class="chapter" data-level="2.9.1" data-path="rmarkdown.html"><a href="rmarkdown.html#html格式"><i class="fa fa-check"></i><b>2.9.1</b> HTML格式</a></li>
<li class="chapter" data-level="2.9.2" data-path="rmarkdown.html"><a href="rmarkdown.html#pdf格式"><i class="fa fa-check"></i><b>2.9.2</b> PDF格式</a></li>
<li class="chapter" data-level="2.9.3" data-path="rmarkdown.html"><a href="rmarkdown.html#word格式"><i class="fa fa-check"></i><b>2.9.3</b> Word格式</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="rmarkdown-beyond.html"><a href="rmarkdown-beyond.html"><i class="fa fa-check"></i><b>3</b> Rmarkdown拓展</a><ul>
<li class="chapter" data-level="3.1" data-path="rmarkdown-beyond.html"><a href="rmarkdown-beyond.html#blogdown"><i class="fa fa-check"></i><b>3.1</b> Blogdown</a><ul>
<li class="chapter" data-level="3.1.1" data-path="rmarkdown-beyond.html"><a href="rmarkdown-beyond.html#需要准备的东西"><i class="fa fa-check"></i><b>3.1.1</b> 需要准备的东西</a></li>
<li class="chapter" data-level="3.1.2" data-path="rmarkdown-beyond.html"><a href="rmarkdown-beyond.html#安装-1"><i class="fa fa-check"></i><b>3.1.2</b> 安装</a></li>
<li class="chapter" data-level="3.1.3" data-path="rmarkdown-beyond.html"><a href="rmarkdown-beyond.html#创建项目"><i class="fa fa-check"></i><b>3.1.3</b> 创建项目</a></li>
<li class="chapter" data-level="3.1.4" data-path="rmarkdown-beyond.html"><a href="rmarkdown-beyond.html#预览"><i class="fa fa-check"></i><b>3.1.4</b> 预览</a></li>
<li class="chapter" data-level="3.1.5" data-path="rmarkdown-beyond.html"><a href="rmarkdown-beyond.html#使用rstudio-ide来管理网站"><i class="fa fa-check"></i><b>3.1.5</b> 使用Rstudio IDE来管理网站</a></li>
<li class="chapter" data-level="3.1.6" data-path="rmarkdown-beyond.html"><a href="rmarkdown-beyond.html#更改为其他的主题"><i class="fa fa-check"></i><b>3.1.6</b> 更改为其他的主题</a></li>
<li class="chapter" data-level="3.1.7" data-path="rmarkdown-beyond.html"><a href="rmarkdown-beyond.html#发布自己的网站"><i class="fa fa-check"></i><b>3.1.7</b> 发布自己的网站</a></li>
<li class="chapter" data-level="3.1.8" data-path="rmarkdown-beyond.html"><a href="rmarkdown-beyond.html#发布到github上"><i class="fa fa-check"></i><b>3.1.8</b> 发布到github上</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="rmarkdown-beyond.html"><a href="rmarkdown-beyond.html#bookdown"><i class="fa fa-check"></i><b>3.2</b> Bookdown</a></li>
<li class="chapter" data-level="3.3" data-path="rmarkdown-beyond.html"><a href="rmarkdown-beyond.html#presentation"><i class="fa fa-check"></i><b>3.3</b> Presentation</a></li>
<li class="chapter" data-level="3.4" data-path="rmarkdown-beyond.html"><a href="rmarkdown-beyond.html#pagedown"><i class="fa fa-check"></i><b>3.4</b> Pagedown</a></li>
<li class="chapter" data-level="3.5" data-path="rmarkdown-beyond.html"><a href="rmarkdown-beyond.html#pkgdown"><i class="fa fa-check"></i><b>3.5</b> Pkgdown</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="git.html"><a href="git.html"><i class="fa fa-check"></i><b>4</b> Git和github</a><ul>
<li class="chapter" data-level="4.1" data-path="git.html"><a href="git.html#安装git"><i class="fa fa-check"></i><b>4.1</b> 安装git</a></li>
<li class="chapter" data-level="4.2" data-path="git.html"><a href="git.html#注册github账户"><i class="fa fa-check"></i><b>4.2</b> 注册github账户</a></li>
<li class="chapter" data-level="4.3" data-path="git.html"><a href="git.html#git基础知识"><i class="fa fa-check"></i><b>4.3</b> Git基础知识</a></li>
<li class="chapter" data-level="4.4" data-path="git.html"><a href="git.html#git配制"><i class="fa fa-check"></i><b>4.4</b> Git配制</a><ul>
<li class="chapter" data-level="4.4.1" data-path="git.html"><a href="git.html#安装git之后需要进行一些全局设置比如用户名邮箱."><i class="fa fa-check"></i><b>4.4.1</b> 安装git之后,需要进行一些全局设置,比如用户名邮箱.</a></li>
<li class="chapter" data-level="4.4.2" data-path="git.html"><a href="git.html#生成秘钥"><i class="fa fa-check"></i><b>4.4.2</b> 生成秘钥</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="git.html"><a href="git.html#git基本操作"><i class="fa fa-check"></i><b>4.5</b> Git基本操作</a><ul>
<li class="chapter" data-level="4.5.1" data-path="git.html"><a href="git.html#创建本地git仓库reop"><i class="fa fa-check"></i><b>4.5.1</b> 创建本地git仓库(reop)</a></li>
<li class="chapter" data-level="4.5.2" data-path="git.html"><a href="git.html#将文件添加到版本库"><i class="fa fa-check"></i><b>4.5.2</b> 将文件添加到版本库</a></li>
<li class="chapter" data-level="4.5.3" data-path="git.html"><a href="git.html#查看仓库状态"><i class="fa fa-check"></i><b>4.5.3</b> 查看仓库状态</a></li>
<li class="chapter" data-level="4.5.4" data-path="git.html"><a href="git.html#查看仓库中的具体修改"><i class="fa fa-check"></i><b>4.5.4</b> 查看仓库中的具体修改</a></li>
<li class="chapter" data-level="4.5.5" data-path="git.html"><a href="git.html#查看提交的历史记录"><i class="fa fa-check"></i><b>4.5.5</b> 查看提交的历史记录</a></li>
<li class="chapter" data-level="4.5.6" data-path="git.html"><a href="git.html#版本回退"><i class="fa fa-check"></i><b>4.5.6</b> 版本回退</a></li>
<li class="chapter" data-level="4.5.7" data-path="git.html"><a href="git.html#回到未来某个版本"><i class="fa fa-check"></i><b>4.5.7</b> 回到未来某个版本</a></li>
<li class="chapter" data-level="4.5.8" data-path="git.html"><a href="git.html#撤销修改"><i class="fa fa-check"></i><b>4.5.8</b> 撤销修改</a></li>
<li class="chapter" data-level="4.5.9" data-path="git.html"><a href="git.html#删除文件"><i class="fa fa-check"></i><b>4.5.9</b> 删除文件</a></li>
</ul></li>
<li class="chapter" data-level="4.6" data-path="git.html"><a href="git.html#git的分支管理"><i class="fa fa-check"></i><b>4.6</b> git的分支管理</a><ul>
<li class="chapter" data-level="4.6.1" data-path="git.html"><a href="git.html#查看分支"><i class="fa fa-check"></i><b>4.6.1</b> 查看分支</a></li>
<li class="chapter" data-level="4.6.2" data-path="git.html"><a href="git.html#创建分支"><i class="fa fa-check"></i><b>4.6.2</b> 创建分支</a></li>
<li class="chapter" data-level="4.6.3" data-path="git.html"><a href="git.html#切换分支"><i class="fa fa-check"></i><b>4.6.3</b> 切换分支</a></li>
<li class="chapter" data-level="4.6.4" data-path="git.html"><a href="git.html#switch命令"><i class="fa fa-check"></i><b>4.6.4</b> <code>switch</code>命令</a></li>
<li class="chapter" data-level="4.6.5" data-path="git.html"><a href="git.html#合并分支merge"><i class="fa fa-check"></i><b>4.6.5</b> 合并分支(merge)</a></li>
<li class="chapter" data-level="4.6.6" data-path="git.html"><a href="git.html#删除分支"><i class="fa fa-check"></i><b>4.6.6</b> 删除分支</a></li>
<li class="chapter" data-level="4.6.7" data-path="git.html"><a href="git.html#分支提交冲突"><i class="fa fa-check"></i><b>4.6.7</b> 分支提交冲突</a></li>
<li class="chapter" data-level="4.6.8" data-path="git.html"><a href="git.html#分支策略"><i class="fa fa-check"></i><b>4.6.8</b> 分支策略</a></li>
</ul></li>
<li class="chapter" data-level="4.7" data-path="git.html"><a href="git.html#远程仓库github"><i class="fa fa-check"></i><b>4.7</b> 远程仓库(GitHub)</a><ul>
<li class="chapter" data-level="4.7.1" data-path="git.html"><a href="git.html#添加到远程仓库"><i class="fa fa-check"></i><b>4.7.1</b> 添加到远程仓库</a></li>
<li class="chapter" data-level="4.7.2" data-path="git.html"><a href="git.html#推送到远程仓库"><i class="fa fa-check"></i><b>4.7.2</b> 推送到远程仓库</a></li>
<li class="chapter" data-level="4.7.3" data-path="git.html"><a href="git.html#从远程仓库获取最新内容"><i class="fa fa-check"></i><b>4.7.3</b> 从远程仓库获取最新内容</a></li>
<li class="chapter" data-level="4.7.4" data-path="git.html"><a href="git.html#查看远程仓库信息"><i class="fa fa-check"></i><b>4.7.4</b> 查看远程仓库信息</a></li>
<li class="chapter" data-level="4.7.5" data-path="git.html"><a href="git.html#从远程仓库克隆"><i class="fa fa-check"></i><b>4.7.5</b> 从远程仓库克隆</a></li>
</ul></li>
<li class="chapter" data-level="4.8" data-path="git.html"><a href="git.html#设置忽略文件"><i class="fa fa-check"></i><b>4.8</b> 设置忽略文件</a></li>
<li class="chapter" data-level="4.9" data-path="git.html"><a href="git.html#多人协作"><i class="fa fa-check"></i><b>4.9</b> 多人协作</a><ul>
<li class="chapter" data-level="4.9.1" data-path="git.html"><a href="git.html#创建远程仓库"><i class="fa fa-check"></i><b>4.9.1</b> 创建远程仓库</a></li>
<li class="chapter" data-level="4.9.2" data-path="git.html"><a href="git.html#本地创建分支并推送到远程仓库"><i class="fa fa-check"></i><b>4.9.2</b> 本地创建分支并推送到远程仓库</a></li>
<li class="chapter" data-level="4.9.3" data-path="git.html"><a href="git.html#邀请合作者"><i class="fa fa-check"></i><b>4.9.3</b> 邀请合作者</a></li>
<li class="chapter" data-level="4.9.4" data-path="git.html"><a href="git.html#合作者参与项目"><i class="fa fa-check"></i><b>4.9.4</b> 合作者参与项目</a></li>
<li class="chapter" data-level="4.9.5" data-path="git.html"><a href="git.html#克隆项目"><i class="fa fa-check"></i><b>4.9.5</b> 克隆项目</a></li>
<li class="chapter" data-level="4.9.6" data-path="git.html"><a href="git.html#开始工作"><i class="fa fa-check"></i><b>4.9.6</b> 开始工作</a></li>
<li class="chapter" data-level="4.9.7" data-path="git.html"><a href="git.html#有冲突怎么办"><i class="fa fa-check"></i><b>4.9.7</b> 有冲突怎么办</a></li>
<li class="chapter" data-level="4.9.8" data-path="git.html"><a href="git.html#抓取分支"><i class="fa fa-check"></i><b>4.9.8</b> 抓取分支</a></li>
<li class="chapter" data-level="4.9.9" data-path="git.html"><a href="git.html#多人协作的工作模式"><i class="fa fa-check"></i><b>4.9.9</b> 多人协作的工作模式</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="ggplot2.html"><a href="ggplot2.html"><i class="fa fa-check"></i><b>5</b> ggplot2</a><ul>
<li class="chapter" data-level="5.1" data-path="ggplot2.html"><a href="ggplot2.html#安装-2"><i class="fa fa-check"></i><b>5.1</b> 安装</a></li>
<li class="chapter" data-level="5.2" data-path="ggplot2.html"><a href="ggplot2.html#基础知识"><i class="fa fa-check"></i><b>5.2</b> 基础知识</a><ul>
<li class="chapter" data-level="5.2.1" data-path="ggplot2.html"><a href="ggplot2.html#ggplot图像的三个元素"><i class="fa fa-check"></i><b>5.2.1</b> ggplot图像的三个元素</a></li>
<li class="chapter" data-level="5.2.2" data-path="ggplot2.html"><a href="ggplot2.html#颜色性状大小和其他的美学属性aesthetic-attributes"><i class="fa fa-check"></i><b>5.2.2</b> 颜色,性状,大小和其他的美学属性(aesthetic attributes)</a></li>
<li class="chapter" data-level="5.2.3" data-path="ggplot2.html"><a href="ggplot2.html#分面facetting"><i class="fa fa-check"></i><b>5.2.3</b> 分面(Facetting)</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="network.html"><a href="network.html"><i class="fa fa-check"></i><b>6</b> Network分析总结</a><ul>
<li class="chapter" data-level="6.1" data-path="network.html"><a href="network.html#example-one"><i class="fa fa-check"></i><b>6.1</b> Example one</a></li>
<li class="chapter" data-level="6.2" data-path="network.html"><a href="network.html#example-two"><i class="fa fa-check"></i><b>6.2</b> Example two</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="multiomics.html"><a href="multiomics.html"><i class="fa fa-check"></i><b>7</b> 多组学分析</a><ul>
<li class="chapter" data-level="7.1" data-path="multiomics.html"><a href="multiomics.html#wgcna"><i class="fa fa-check"></i><b>7.1</b> WGCNA</a><ul>
<li class="chapter" data-level="7.1.1" data-path="multiomics.html"><a href="multiomics.html#data-input-and-cleaning"><i class="fa fa-check"></i><b>7.1.1</b> Data input and cleaning</a></li>
<li class="chapter" data-level="7.1.2" data-path="multiomics.html"><a href="multiomics.html#network-analysis-of-liver-expression-data-in-female-mice"><i class="fa fa-check"></i><b>7.1.2</b> Network analysis of liver expression data in female mice</a></li>
<li class="chapter" data-level="7.1.3" data-path="multiomics.html"><a href="multiomics.html#relating-modules-to-external-information-and-identifying-important"><i class="fa fa-check"></i><b>7.1.3</b> Relating modules to external information and identifying important</a></li>
<li class="chapter" data-level="7.1.4" data-path="multiomics.html"><a href="multiomics.html#对module进行分析"><i class="fa fa-check"></i><b>7.1.4</b> 对module进行分析</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="regularexpression.html"><a href="regularexpression.html"><i class="fa fa-check"></i><b>8</b> 正则表达式</a><ul>
<li class="chapter" data-level="8.1" data-path="regularexpression.html"><a href="regularexpression.html#特殊字符"><i class="fa fa-check"></i><b>8.1</b> 特殊字符</a></li>
<li class="chapter" data-level="8.2" data-path="regularexpression.html"><a href="regularexpression.html#匹配match"><i class="fa fa-check"></i><b>8.2</b> 匹配(match)</a></li>
<li class="chapter" data-level="8.3" data-path="regularexpression.html"><a href="regularexpression.html#替换-alternates"><i class="fa fa-check"></i><b>8.3</b> 替换 (Alternates)</a></li>
<li class="chapter" data-level="8.4" data-path="regularexpression.html"><a href="regularexpression.html#锚点anchors"><i class="fa fa-check"></i><b>8.4</b> 锚点(anchors)</a><ul>
<li class="chapter" data-level="8.4.1" data-path="regularexpression.html"><a href="regularexpression.html#以固定要求开头"><i class="fa fa-check"></i><b>8.4.1</b> 以固定要求开头</a></li>
<li class="chapter" data-level="8.4.2" data-path="regularexpression.html"><a href="regularexpression.html#以固定要求结尾"><i class="fa fa-check"></i><b>8.4.2</b> 以固定要求结尾</a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="regularexpression.html"><a href="regularexpression.html#数量quantifiers"><i class="fa fa-check"></i><b>8.5</b> 数量(Quantifiers)</a></li>
<li class="chapter" data-level="8.6" data-path="regularexpression.html"><a href="regularexpression.html#整体匹配-groups"><i class="fa fa-check"></i><b>8.6</b> 整体匹配 (groups)</a></li>
<li class="chapter" data-level="8.7" data-path="regularexpression.html"><a href="regularexpression.html#反义"><i class="fa fa-check"></i><b>8.7</b> 反义</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="tidyverse-stringr.html"><a href="tidyverse-stringr.html"><i class="fa fa-check"></i><b>9</b> <code>Tidyverse</code>用于数据分析stringr</a><ul>
<li class="chapter" data-level="9.1" data-path="tidyverse-stringr.html"><a href="tidyverse-stringr.html#安装-3"><i class="fa fa-check"></i><b>9.1</b> 安装</a></li>
<li class="chapter" data-level="9.2" data-path="tidyverse-stringr.html"><a href="tidyverse-stringr.html#rstudio-addin"><i class="fa fa-check"></i><b>9.2</b> Rstudio Addin</a></li>
<li class="chapter" data-level="9.3" data-path="tidyverse-stringr.html"><a href="tidyverse-stringr.html#pattern-matching"><i class="fa fa-check"></i><b>9.3</b> Pattern matching</a><ul>
<li class="chapter" data-level="9.3.1" data-path="tidyverse-stringr.html"><a href="tidyverse-stringr.html#计算一个文本中符合要求的数目"><i class="fa fa-check"></i><b>9.3.1</b> 计算一个文本中符合要求的数目</a></li>
<li class="chapter" data-level="9.3.2" data-path="tidyverse-stringr.html"><a href="tidyverse-stringr.html#判断一个文本中是否存在一个pattern"><i class="fa fa-check"></i><b>9.3.2</b> 判断一个文本中是否存在一个pattern</a></li>
<li class="chapter" data-level="9.3.3" data-path="tidyverse-stringr.html"><a href="tidyverse-stringr.html#从文本中提取pattern"><i class="fa fa-check"></i><b>9.3.3</b> 从文本中提取pattern</a></li>
<li class="chapter" data-level="9.3.4" data-path="tidyverse-stringr.html"><a href="tidyverse-stringr.html#确定某个pattern在文本中的位置"><i class="fa fa-check"></i><b>9.3.4</b> 确定某个pattern在文本中的位置</a></li>
<li class="chapter" data-level="9.3.5" data-path="tidyverse-stringr.html"><a href="tidyverse-stringr.html#从文本中提取extract匹配到的group"><i class="fa fa-check"></i><b>9.3.5</b> 从文本中提取(extract)匹配到的group</a></li>
<li class="chapter" data-level="9.3.6" data-path="tidyverse-stringr.html"><a href="tidyverse-stringr.html#从文本中去除匹配到的pattern"><i class="fa fa-check"></i><b>9.3.6</b> 从文本中去除匹配到的pattern</a></li>
<li class="chapter" data-level="9.3.7" data-path="tidyverse-stringr.html"><a href="tidyverse-stringr.html#从文本中替换匹配到的pattern"><i class="fa fa-check"></i><b>9.3.7</b> 从文本中替换匹配到的pattern</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="biostatistics.html"><a href="biostatistics.html"><i class="fa fa-check"></i><b>10</b> Modern Statistics for Modern Biology</a></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://shenxt.info" target="blank">Find me in my blog</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">R cookbook from Xiaotao Shen</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="multiomics" class="section level1">
<h1><span class="header-section-number">Chapter 7</span> 多组学分析</h1>
<div id="wgcna" class="section level2">
<h2><span class="header-section-number">7.1</span> WGCNA</h2>
<p>使用官方例子进行学习.</p>
<p><a href="https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/Tutorials/" class="uri">https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/Tutorials/</a></p>
<div id="data-input-and-cleaning" class="section level3">
<h3><span class="header-section-number">7.1.1</span> Data input and cleaning</h3>
<ol style="list-style-type: decimal">
<li>首先读取数据</li>
</ol>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw">library</span>(WGCNA)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2">femData &lt;-<span class="st"> </span></a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="st">  </span><span class="kw">read.csv</span>(<span class="st">&quot;/Users/shenxt/study/WGCNA_test/LiverFemale3600.csv&quot;</span>)</a>
<a class="sourceLine" id="cb1-4" data-line-number="4"><span class="kw">dim</span>(femData)</a></code></pre></div>
<pre><code>## [1] 3600  143</code></pre>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="kw">colnames</span>(femData)</a></code></pre></div>
<pre><code>##   [1] &quot;substanceBXH&quot;   &quot;gene_symbol&quot;    &quot;LocusLinkID&quot;    &quot;ProteomeID&quot;     &quot;cytogeneticLoc&quot; &quot;CHROMOSOME&quot;    
##   [7] &quot;StartPosition&quot;  &quot;EndPosition&quot;    &quot;F2_2&quot;           &quot;F2_3&quot;           &quot;F2_14&quot;          &quot;F2_15&quot;         
##  [13] &quot;F2_19&quot;          &quot;F2_20&quot;          &quot;F2_23&quot;          &quot;F2_24&quot;          &quot;F2_26&quot;          &quot;F2_37&quot;         
##  [19] &quot;F2_42&quot;          &quot;F2_43&quot;          &quot;F2_45&quot;          &quot;F2_46&quot;          &quot;F2_47&quot;          &quot;F2_48&quot;         
##  [25] &quot;F2_51&quot;          &quot;F2_52&quot;          &quot;F2_54&quot;          &quot;F2_63&quot;          &quot;F2_65&quot;          &quot;F2_66&quot;         
##  [31] &quot;F2_68&quot;          &quot;F2_69&quot;          &quot;F2_70&quot;          &quot;F2_71&quot;          &quot;F2_72&quot;          &quot;F2_78&quot;         
##  [37] &quot;F2_79&quot;          &quot;F2_80&quot;          &quot;F2_81&quot;          &quot;F2_83&quot;          &quot;F2_86&quot;          &quot;F2_87&quot;         
##  [43] &quot;F2_88&quot;          &quot;F2_89&quot;          &quot;F2_107&quot;         &quot;F2_108&quot;         &quot;F2_109&quot;         &quot;F2_110&quot;        
##  [49] &quot;F2_111&quot;         &quot;F2_112&quot;         &quot;F2_117&quot;         &quot;F2_119&quot;         &quot;F2_125&quot;         &quot;F2_126&quot;        
##  [55] &quot;F2_127&quot;         &quot;F2_141&quot;         &quot;F2_142&quot;         &quot;F2_143&quot;         &quot;F2_144&quot;         &quot;F2_145&quot;        
##  [61] &quot;F2_154&quot;         &quot;F2_155&quot;         &quot;F2_156&quot;         &quot;F2_157&quot;         &quot;F2_162&quot;         &quot;F2_163&quot;        
##  [67] &quot;F2_164&quot;         &quot;F2_165&quot;         &quot;F2_166&quot;         &quot;F2_167&quot;         &quot;F2_169&quot;         &quot;F2_180&quot;        
##  [73] &quot;F2_181&quot;         &quot;F2_182&quot;         &quot;F2_187&quot;         &quot;F2_188&quot;         &quot;F2_189&quot;         &quot;F2_190&quot;        
##  [79] &quot;F2_191&quot;         &quot;F2_192&quot;         &quot;F2_194&quot;         &quot;F2_195&quot;         &quot;F2_200&quot;         &quot;F2_201&quot;        
##  [85] &quot;F2_212&quot;         &quot;F2_213&quot;         &quot;F2_214&quot;         &quot;F2_215&quot;         &quot;F2_221&quot;         &quot;F2_222&quot;        
##  [91] &quot;F2_223&quot;         &quot;F2_224&quot;         &quot;F2_225&quot;         &quot;F2_226&quot;         &quot;F2_227&quot;         &quot;F2_228&quot;        
##  [97] &quot;F2_241&quot;         &quot;F2_242&quot;         &quot;F2_243&quot;         &quot;F2_244&quot;         &quot;F2_245&quot;         &quot;F2_247&quot;        
## [103] &quot;F2_248&quot;         &quot;F2_261&quot;         &quot;F2_263&quot;         &quot;F2_264&quot;         &quot;F2_270&quot;         &quot;F2_271&quot;        
## [109] &quot;F2_272&quot;         &quot;F2_278&quot;         &quot;F2_287&quot;         &quot;F2_288&quot;         &quot;F2_289&quot;         &quot;F2_290&quot;        
## [115] &quot;F2_291&quot;         &quot;F2_296&quot;         &quot;F2_298&quot;         &quot;F2_299&quot;         &quot;F2_300&quot;         &quot;F2_302&quot;        
## [121] &quot;F2_303&quot;         &quot;F2_304&quot;         &quot;F2_305&quot;         &quot;F2_306&quot;         &quot;F2_307&quot;         &quot;F2_308&quot;        
## [127] &quot;F2_309&quot;         &quot;F2_310&quot;         &quot;F2_311&quot;         &quot;F2_312&quot;         &quot;F2_320&quot;         &quot;F2_321&quot;        
## [133] &quot;F2_323&quot;         &quot;F2_324&quot;         &quot;F2_325&quot;         &quot;F2_326&quot;         &quot;F2_327&quot;         &quot;F2_328&quot;        
## [139] &quot;F2_329&quot;         &quot;F2_330&quot;         &quot;F2_332&quot;         &quot;F2_355&quot;         &quot;F2_357&quot;</code></pre>
<p>可以看到该数据行是基因,列是样品.</p>
<p>然后将基因的描述列去掉.并且将其转置,行是样品,列是基因.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1">data_exp0 &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">t</span>(femData[, <span class="op">-</span><span class="kw">c</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">8</span>)]), <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb5-2" data-line-number="2"><span class="kw">names</span>(data_exp0) =<span class="st"> </span>femData<span class="op">$</span>substanceBXH</a>
<a class="sourceLine" id="cb5-3" data-line-number="3"><span class="kw">rownames</span>(data_exp0) =<span class="st"> </span><span class="kw">names</span>(femData)[<span class="op">-</span><span class="kw">c</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">8</span>)]</a></code></pre></div>
<p>然后需要去除掉数据(表达矩阵)中的outlier基因和样品</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1">gsg &lt;-<span class="st"> </span><span class="kw">goodSamplesGenes</span>(data_exp0, <span class="dt">verbose =</span> <span class="dv">3</span>)</a></code></pre></div>
<pre><code>##  Flagging genes and samples with too many missing values...
##   ..step 1</code></pre>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1">gsg<span class="op">$</span>allOK</a></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<p>该函数用来检查数据中的基因是否符合要求,如果<code>allOK</code>为TRUE,则代表通过检查,如果不是TRUE,则需要手动去除基因和样品:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1"><span class="cf">if</span> (<span class="op">!</span>gsg<span class="op">$</span>allOK)</a>
<a class="sourceLine" id="cb10-2" data-line-number="2">{</a>
<a class="sourceLine" id="cb10-3" data-line-number="3">  <span class="co"># Optionally, print the gene and sample names that were removed:</span></a>
<a class="sourceLine" id="cb10-4" data-line-number="4">  <span class="cf">if</span> (<span class="kw">sum</span>(<span class="op">!</span>gsg<span class="op">$</span>goodGenes) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>)</a>
<a class="sourceLine" id="cb10-5" data-line-number="5">    <span class="kw">printFlush</span>(<span class="kw">paste</span>(<span class="st">&quot;Removing genes:&quot;</span>, <span class="kw">paste</span>(<span class="kw">names</span>(data_exp0)[<span class="op">!</span>gsg<span class="op">$</span>goodGenes], <span class="dt">collapse =</span> <span class="st">&quot;, &quot;</span>)))</a>
<a class="sourceLine" id="cb10-6" data-line-number="6">  </a>
<a class="sourceLine" id="cb10-7" data-line-number="7">  <span class="cf">if</span> (<span class="kw">sum</span>(<span class="op">!</span>gsg<span class="op">$</span>goodSamples) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>)</a>
<a class="sourceLine" id="cb10-8" data-line-number="8">    <span class="kw">printFlush</span>(<span class="kw">paste</span>(<span class="st">&quot;Removing samples:&quot;</span>, <span class="kw">paste</span>(<span class="kw">rownames</span>(data_exp0)[<span class="op">!</span>gsg<span class="op">$</span>goodSamples], <span class="dt">collapse =</span> <span class="st">&quot;, &quot;</span>)))</a>
<a class="sourceLine" id="cb10-9" data-line-number="9">  </a>
<a class="sourceLine" id="cb10-10" data-line-number="10">  <span class="co"># Remove the offending genes and samples from the data:</span></a>
<a class="sourceLine" id="cb10-11" data-line-number="11">  data_exp0 &lt;-<span class="st"> </span>data_exp0[gsg<span class="op">$</span>goodSamples, gsg<span class="op">$</span>goodGenes]</a>
<a class="sourceLine" id="cb10-12" data-line-number="12">}</a></code></pre></div>
<p>现在下面对样品进行聚类,然后观察有哪些明显的outlier samples.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1">sampleTree &lt;-<span class="st"> </span><span class="kw">hclust</span>(<span class="kw">dist</span>(data_exp0), <span class="dt">method =</span> <span class="st">&quot;average&quot;</span>)</a>
<a class="sourceLine" id="cb11-2" data-line-number="2"><span class="co"># Plot the sample tree: Open a graphic output window of size 12 by 9 inches</span></a>
<a class="sourceLine" id="cb11-3" data-line-number="3"><span class="co"># The user should change the dimensions if the window is too large or too small.</span></a>
<a class="sourceLine" id="cb11-4" data-line-number="4"><span class="kw">sizeGrWindow</span>(<span class="dv">12</span>, <span class="dv">9</span>)</a>
<a class="sourceLine" id="cb11-5" data-line-number="5"><span class="co">#pdf(file = &quot;Plots/sampleClustering.pdf&quot;, width = 12, height = 9);</span></a>
<a class="sourceLine" id="cb11-6" data-line-number="6"><span class="kw">par</span>(<span class="dt">cex =</span> <span class="fl">0.6</span>)</a>
<a class="sourceLine" id="cb11-7" data-line-number="7"><span class="kw">par</span>(<span class="dt">mar =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">4</span>, <span class="dv">2</span>, <span class="dv">0</span>))</a>
<a class="sourceLine" id="cb11-8" data-line-number="8"><span class="kw">plot</span>(</a>
<a class="sourceLine" id="cb11-9" data-line-number="9">  sampleTree,</a>
<a class="sourceLine" id="cb11-10" data-line-number="10">  <span class="dt">main =</span> <span class="st">&quot;Sample clustering to detect outliers&quot;</span>,</a>
<a class="sourceLine" id="cb11-11" data-line-number="11">  <span class="dt">sub =</span> <span class="st">&quot;&quot;</span>,</a>
<a class="sourceLine" id="cb11-12" data-line-number="12">  <span class="dt">xlab =</span> <span class="st">&quot;&quot;</span>,</a>
<a class="sourceLine" id="cb11-13" data-line-number="13">  <span class="dt">cex.lab =</span> <span class="fl">1.5</span>,</a>
<a class="sourceLine" id="cb11-14" data-line-number="14">  <span class="dt">cex.axis =</span> <span class="fl">1.5</span>,</a>
<a class="sourceLine" id="cb11-15" data-line-number="15">  <span class="dt">cex.main =</span> <span class="dv">2</span></a>
<a class="sourceLine" id="cb11-16" data-line-number="16">)</a>
<a class="sourceLine" id="cb11-17" data-line-number="17"></a>
<a class="sourceLine" id="cb11-18" data-line-number="18"><span class="co"># Plot a line to show the cut</span></a>
<a class="sourceLine" id="cb11-19" data-line-number="19"><span class="kw">abline</span>(<span class="dt">h =</span> <span class="dv">15</span>, <span class="dt">col =</span> <span class="st">&quot;red&quot;</span>)</a></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1"><span class="co"># Determine cluster under the line</span></a>
<a class="sourceLine" id="cb12-2" data-line-number="2">clust =<span class="st"> </span><span class="kw">cutreeStatic</span>(sampleTree, <span class="dt">cutHeight =</span> <span class="dv">15</span>, <span class="dt">minSize =</span> <span class="dv">10</span>)</a>
<a class="sourceLine" id="cb12-3" data-line-number="3"><span class="kw">table</span>(clust)</a></code></pre></div>
<pre><code>## clust
##   0   1 
##   1 134</code></pre>
<p>可以明显的看到一个outlier sample.可以手动删除,也可以自动的办法.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1"><span class="co"># clust 1 contains the samples we want to keep.</span></a>
<a class="sourceLine" id="cb14-2" data-line-number="2">keepSamples =<span class="st"> </span>(clust <span class="op">==</span><span class="st"> </span><span class="dv">1</span>)</a>
<a class="sourceLine" id="cb14-3" data-line-number="3">datExpr =<span class="st"> </span>data_exp0[keepSamples,]</a>
<a class="sourceLine" id="cb14-4" data-line-number="4">nGenes =<span class="st"> </span><span class="kw">ncol</span>(datExpr)</a>
<a class="sourceLine" id="cb14-5" data-line-number="5">nSamples =<span class="st"> </span><span class="kw">nrow</span>(datExpr)</a></code></pre></div>
<p><code>datExpr</code>就是我们最终用来进行network分析的表达数据.</p>
<p>下面我们需要读取clinical data.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1">traitData =<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;/Users/shenxt/study/WGCNA_test/ClinicalTraits.csv&quot;</span>)</a>
<a class="sourceLine" id="cb15-2" data-line-number="2"><span class="kw">dim</span>(traitData)</a></code></pre></div>
<pre><code>## [1] 361  38</code></pre>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1"><span class="kw">names</span>(traitData)</a></code></pre></div>
<pre><code>##  [1] &quot;X&quot;                  &quot;Mice&quot;               &quot;Number&quot;             &quot;Mouse_ID&quot;           &quot;Strain&quot;            
##  [6] &quot;sex&quot;                &quot;DOB&quot;                &quot;parents&quot;            &quot;Western_Diet&quot;       &quot;Sac_Date&quot;          
## [11] &quot;weight_g&quot;           &quot;length_cm&quot;          &quot;ab_fat&quot;             &quot;other_fat&quot;          &quot;total_fat&quot;         
## [16] &quot;comments&quot;           &quot;X100xfat_weight&quot;    &quot;Trigly&quot;             &quot;Total_Chol&quot;         &quot;HDL_Chol&quot;          
## [21] &quot;UC&quot;                 &quot;FFA&quot;                &quot;Glucose&quot;            &quot;LDL_plus_VLDL&quot;      &quot;MCP_1_phys&quot;        
## [26] &quot;Insulin_ug_l&quot;       &quot;Glucose_Insulin&quot;    &quot;Leptin_pg_ml&quot;       &quot;Adiponectin&quot;        &quot;Aortic.lesions&quot;    
## [31] &quot;Note&quot;               &quot;Aneurysm&quot;           &quot;Aortic_cal_M&quot;       &quot;Aortic_cal_L&quot;       &quot;CoronaryArtery_Cal&quot;
## [36] &quot;Myocardial_cal&quot;     &quot;BMD_all_limbs&quot;      &quot;BMD_femurs_only&quot;</code></pre>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1"><span class="co"># remove columns that hold information we do not need.</span></a>
<a class="sourceLine" id="cb19-2" data-line-number="2">allTraits =<span class="st"> </span>traitData[, <span class="op">-</span><span class="kw">c</span>(<span class="dv">31</span>, <span class="dv">16</span>)]</a>
<a class="sourceLine" id="cb19-3" data-line-number="3">allTraits =<span class="st"> </span>allTraits[, <span class="kw">c</span>(<span class="dv">2</span>, <span class="dv">11</span><span class="op">:</span><span class="dv">36</span>)]</a>
<a class="sourceLine" id="cb19-4" data-line-number="4"><span class="kw">dim</span>(allTraits)</a></code></pre></div>
<pre><code>## [1] 361  27</code></pre>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" data-line-number="1"><span class="kw">names</span>(allTraits)</a></code></pre></div>
<pre><code>##  [1] &quot;Mice&quot;               &quot;weight_g&quot;           &quot;length_cm&quot;          &quot;ab_fat&quot;             &quot;other_fat&quot;         
##  [6] &quot;total_fat&quot;          &quot;X100xfat_weight&quot;    &quot;Trigly&quot;             &quot;Total_Chol&quot;         &quot;HDL_Chol&quot;          
## [11] &quot;UC&quot;                 &quot;FFA&quot;                &quot;Glucose&quot;            &quot;LDL_plus_VLDL&quot;      &quot;MCP_1_phys&quot;        
## [16] &quot;Insulin_ug_l&quot;       &quot;Glucose_Insulin&quot;    &quot;Leptin_pg_ml&quot;       &quot;Adiponectin&quot;        &quot;Aortic.lesions&quot;    
## [21] &quot;Aneurysm&quot;           &quot;Aortic_cal_M&quot;       &quot;Aortic_cal_L&quot;       &quot;CoronaryArtery_Cal&quot; &quot;Myocardial_cal&quot;    
## [26] &quot;BMD_all_limbs&quot;      &quot;BMD_femurs_only&quot;</code></pre>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" data-line-number="1"><span class="co"># Form a data frame analogous to expression data that will hold the clinical traits.</span></a>
<a class="sourceLine" id="cb23-2" data-line-number="2">femaleSamples =<span class="st"> </span><span class="kw">rownames</span>(datExpr)</a>
<a class="sourceLine" id="cb23-3" data-line-number="3">traitRows =<span class="st"> </span><span class="kw">match</span>(femaleSamples, allTraits<span class="op">$</span>Mice)</a>
<a class="sourceLine" id="cb23-4" data-line-number="4">datTraits =<span class="st"> </span>allTraits[traitRows, <span class="dv">-1</span>]</a>
<a class="sourceLine" id="cb23-5" data-line-number="5"><span class="kw">rownames</span>(datTraits) =<span class="st"> </span>allTraits[traitRows, <span class="dv">1</span>]</a>
<a class="sourceLine" id="cb23-6" data-line-number="6"><span class="kw">collectGarbage</span>()</a></code></pre></div>
<p>这时候我们得到的<code>datTraits</code>就是clinicaldata,其中行为样品,列为变量.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb24-1" data-line-number="1"><span class="kw">rownames</span>(datTraits) <span class="op">==</span><span class="st"> </span><span class="kw">rownames</span>(datExpr)</a></code></pre></div>
<pre><code>##   [1] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE
##  [24] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE
##  [47] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE
##  [70] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE
##  [93] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE
## [116] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE</code></pre>
<p>下面我们来看看样品的临床数据的热图和样品的基因的聚类.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb26-1" data-line-number="1"><span class="co"># Re-cluster samples</span></a>
<a class="sourceLine" id="cb26-2" data-line-number="2">sampleTree2 =<span class="st"> </span><span class="kw">hclust</span>(<span class="kw">dist</span>(datExpr), <span class="dt">method =</span> <span class="st">&quot;average&quot;</span>)</a>
<a class="sourceLine" id="cb26-3" data-line-number="3"><span class="co"># Convert traits to a color representation: white means low, red means high, grey means missing entry</span></a>
<a class="sourceLine" id="cb26-4" data-line-number="4">traitColors =<span class="st"> </span><span class="kw">numbers2colors</span>(datTraits, <span class="dt">signed =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb26-5" data-line-number="5"></a>
<a class="sourceLine" id="cb26-6" data-line-number="6"><span class="co"># Plot the sample dendrogram and the colors underneath.</span></a>
<a class="sourceLine" id="cb26-7" data-line-number="7"><span class="kw">plotDendroAndColors</span>(sampleTree2,</a>
<a class="sourceLine" id="cb26-8" data-line-number="8">                    traitColors,</a>
<a class="sourceLine" id="cb26-9" data-line-number="9">                    <span class="dt">groupLabels =</span> <span class="kw">names</span>(datTraits),</a>
<a class="sourceLine" id="cb26-10" data-line-number="10">                    <span class="dt">main =</span> <span class="st">&quot;Sample dendrogram and trait heatmap&quot;</span>)</a></code></pre></div>
<p><img src="r_cookbook_files/figure-html/unnamed-chunk-11-1.png" width="100%" /></p>
<p>最后,把数据保存下来,准备后续的分析.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb27-1" data-line-number="1"><span class="kw">save</span>(datExpr, datTraits, <span class="dt">file =</span> <span class="st">&quot;/Users/shenxt/study/WGCNA_test/FemaleLiver-01-dataInput.RData&quot;</span>)</a></code></pre></div>
</div>
<div id="network-analysis-of-liver-expression-data-in-female-mice" class="section level3">
<h3><span class="header-section-number">7.1.2</span> Network analysis of liver expression data in female mice</h3>
<p>数据都准备好之后,下面就是做网络分析和module detection.</p>
<ol style="list-style-type: decimal">
<li>首先需要设置R运行环境,主要是要能够进行多线程处理,然后读取数据:</li>
</ol>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb28-1" data-line-number="1"><span class="kw">enableWGCNAThreads</span>(<span class="dt">nThreads =</span> <span class="dv">6</span>)</a></code></pre></div>
<pre><code>## Allowing parallel execution with up to 6 working processes.</code></pre>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb30-1" data-line-number="1"><span class="co"># Load the data saved in the first part</span></a>
<a class="sourceLine" id="cb30-2" data-line-number="2">lnames =<span class="st"> </span><span class="kw">load</span>(<span class="dt">file =</span> <span class="st">&quot;/Users/shenxt/study/WGCNA_test/FemaleLiver-01-dataInput.RData&quot;</span>)</a>
<a class="sourceLine" id="cb30-3" data-line-number="3"><span class="co">#The variable lnames contains the names of loaded variables.</span></a>
<a class="sourceLine" id="cb30-4" data-line-number="4">lnames</a></code></pre></div>
<pre><code>## [1] &quot;datExpr&quot;   &quot;datTraits&quot;</code></pre>
<ol start="2" style="list-style-type: decimal">
<li>Automatic network construction and module detection</li>
</ol>
<p>首先需要选择合适的<code>soft-thresholding power</code></p>
<p>可以使用<code>pickSoftThreshold()</code>函数.一般需要选择一系列candidate powers,然后根据结果进行选择.</p>
<blockquote>
<p>注意,这段代码在Rstudio中不能运行,运行出错,需要到R自带IDE或者终端中运行.</p>
</blockquote>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb32-1" data-line-number="1"><span class="co"># Choose a set of soft-thresholding powers</span></a>
<a class="sourceLine" id="cb32-2" data-line-number="2">powers =<span class="st"> </span><span class="kw">c</span>(<span class="kw">c</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>), <span class="kw">seq</span>(<span class="dt">from =</span> <span class="dv">12</span>, <span class="dt">to =</span> <span class="dv">20</span>, <span class="dt">by =</span> <span class="dv">2</span>))</a>
<a class="sourceLine" id="cb32-3" data-line-number="3"><span class="co"># Call the network topology analysis function</span></a>
<a class="sourceLine" id="cb32-4" data-line-number="4">sft =<span class="st"> </span><span class="kw">pickSoftThreshold</span>(datExpr, <span class="dt">powerVector =</span> powers, <span class="dt">verbose =</span> <span class="dv">5</span>)</a>
<a class="sourceLine" id="cb32-5" data-line-number="5"><span class="kw">save</span>(sft, <span class="dt">file =</span> <span class="st">&quot;/Users/shenxt/study/WGCNA_test/sft&quot;</span>)</a></code></pre></div>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb33-1" data-line-number="1"><span class="kw">load</span>(<span class="st">&quot;/Users/shenxt/study/WGCNA_test/sft&quot;</span>)</a>
<a class="sourceLine" id="cb33-2" data-line-number="2">powers =<span class="st"> </span><span class="kw">c</span>(<span class="kw">c</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>), <span class="kw">seq</span>(<span class="dt">from =</span> <span class="dv">12</span>, <span class="dt">to =</span> <span class="dv">20</span>, <span class="dt">by =</span> <span class="dv">2</span>))</a>
<a class="sourceLine" id="cb33-3" data-line-number="3"><span class="co"># Plot the results:</span></a>
<a class="sourceLine" id="cb33-4" data-line-number="4"><span class="kw">sizeGrWindow</span>(<span class="dv">9</span>, <span class="dv">5</span>)</a>
<a class="sourceLine" id="cb33-5" data-line-number="5"><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>))</a>
<a class="sourceLine" id="cb33-6" data-line-number="6"></a>
<a class="sourceLine" id="cb33-7" data-line-number="7">cex1 =<span class="st"> </span><span class="fl">0.9</span></a>
<a class="sourceLine" id="cb33-8" data-line-number="8"></a>
<a class="sourceLine" id="cb33-9" data-line-number="9"><span class="co"># Scale-free topology fit index as a function of the soft-thresholding power</span></a>
<a class="sourceLine" id="cb33-10" data-line-number="10"><span class="kw">plot</span>(</a>
<a class="sourceLine" id="cb33-11" data-line-number="11">  sft<span class="op">$</span>fitIndices[, <span class="dv">1</span>],</a>
<a class="sourceLine" id="cb33-12" data-line-number="12">  <span class="op">-</span><span class="kw">sign</span>(sft<span class="op">$</span>fitIndices[, <span class="dv">3</span>]) <span class="op">*</span><span class="st"> </span>sft<span class="op">$</span>fitIndices[, <span class="dv">2</span>],</a>
<a class="sourceLine" id="cb33-13" data-line-number="13">  <span class="dt">xlab =</span> <span class="st">&quot;Soft Threshold (power)&quot;</span>,</a>
<a class="sourceLine" id="cb33-14" data-line-number="14">  <span class="dt">ylab =</span> <span class="st">&quot;Scale Free Topology Model Fit,signed R^2&quot;</span>,</a>
<a class="sourceLine" id="cb33-15" data-line-number="15">  <span class="dt">type =</span> <span class="st">&quot;n&quot;</span>,</a>
<a class="sourceLine" id="cb33-16" data-line-number="16">  <span class="dt">main =</span> <span class="kw">paste</span>(<span class="st">&quot;Scale independence&quot;</span>)</a>
<a class="sourceLine" id="cb33-17" data-line-number="17">)</a>
<a class="sourceLine" id="cb33-18" data-line-number="18"></a>
<a class="sourceLine" id="cb33-19" data-line-number="19"><span class="kw">text</span>(</a>
<a class="sourceLine" id="cb33-20" data-line-number="20">  sft<span class="op">$</span>fitIndices[, <span class="dv">1</span>],</a>
<a class="sourceLine" id="cb33-21" data-line-number="21">  <span class="op">-</span><span class="kw">sign</span>(sft<span class="op">$</span>fitIndices[, <span class="dv">3</span>]) <span class="op">*</span><span class="st"> </span>sft<span class="op">$</span>fitIndices[, <span class="dv">2</span>],</a>
<a class="sourceLine" id="cb33-22" data-line-number="22">  <span class="dt">labels =</span> powers,</a>
<a class="sourceLine" id="cb33-23" data-line-number="23">  <span class="dt">cex =</span> cex1,</a>
<a class="sourceLine" id="cb33-24" data-line-number="24">  <span class="dt">col =</span> <span class="st">&quot;red&quot;</span></a>
<a class="sourceLine" id="cb33-25" data-line-number="25">)</a>
<a class="sourceLine" id="cb33-26" data-line-number="26"></a>
<a class="sourceLine" id="cb33-27" data-line-number="27"><span class="co"># this line corresponds to using an R^2 cut-off of h</span></a>
<a class="sourceLine" id="cb33-28" data-line-number="28"><span class="kw">abline</span>(<span class="dt">h =</span> <span class="fl">0.90</span>, <span class="dt">col =</span> <span class="st">&quot;red&quot;</span>)</a>
<a class="sourceLine" id="cb33-29" data-line-number="29"><span class="co"># Mean connectivity as a function of the soft-thresholding power</span></a>
<a class="sourceLine" id="cb33-30" data-line-number="30"><span class="kw">plot</span>(</a>
<a class="sourceLine" id="cb33-31" data-line-number="31">  sft<span class="op">$</span>fitIndices[, <span class="dv">1</span>],</a>
<a class="sourceLine" id="cb33-32" data-line-number="32">  sft<span class="op">$</span>fitIndices[, <span class="dv">5</span>],</a>
<a class="sourceLine" id="cb33-33" data-line-number="33">  <span class="dt">xlab =</span> <span class="st">&quot;Soft Threshold (power)&quot;</span>,</a>
<a class="sourceLine" id="cb33-34" data-line-number="34">  <span class="dt">ylab =</span> <span class="st">&quot;Mean Connectivity&quot;</span>,</a>
<a class="sourceLine" id="cb33-35" data-line-number="35">  <span class="dt">type =</span> <span class="st">&quot;n&quot;</span>,</a>
<a class="sourceLine" id="cb33-36" data-line-number="36">  <span class="dt">main =</span> <span class="kw">paste</span>(<span class="st">&quot;Mean connectivity&quot;</span>)</a>
<a class="sourceLine" id="cb33-37" data-line-number="37">)</a>
<a class="sourceLine" id="cb33-38" data-line-number="38"><span class="kw">text</span>(</a>
<a class="sourceLine" id="cb33-39" data-line-number="39">  sft<span class="op">$</span>fitIndices[, <span class="dv">1</span>],</a>
<a class="sourceLine" id="cb33-40" data-line-number="40">  sft<span class="op">$</span>fitIndices[, <span class="dv">5</span>],</a>
<a class="sourceLine" id="cb33-41" data-line-number="41">  <span class="dt">labels =</span> powers,</a>
<a class="sourceLine" id="cb33-42" data-line-number="42">  <span class="dt">cex =</span> cex1,</a>
<a class="sourceLine" id="cb33-43" data-line-number="43">  <span class="dt">col =</span> <span class="st">&quot;red&quot;</span></a>
<a class="sourceLine" id="cb33-44" data-line-number="44">)</a></code></pre></div>
<p>这里我们选择6.</p>
<p>然后使用<code>blockwiseModules()</code>函数就可以得到modules.</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb34-1" data-line-number="1">net &lt;-<span class="st"> </span><span class="kw">blockwiseModules</span>(</a>
<a class="sourceLine" id="cb34-2" data-line-number="2">  datExpr,</a>
<a class="sourceLine" id="cb34-3" data-line-number="3">  <span class="dt">power =</span> <span class="dv">6</span>,</a>
<a class="sourceLine" id="cb34-4" data-line-number="4">  <span class="dt">TOMType =</span> <span class="st">&quot;unsigned&quot;</span>,</a>
<a class="sourceLine" id="cb34-5" data-line-number="5">  <span class="dt">minModuleSize =</span> <span class="dv">30</span>,</a>
<a class="sourceLine" id="cb34-6" data-line-number="6">  <span class="dt">reassignThreshold =</span> <span class="dv">0</span>,</a>
<a class="sourceLine" id="cb34-7" data-line-number="7">  <span class="dt">mergeCutHeight =</span> <span class="fl">0.25</span>,</a>
<a class="sourceLine" id="cb34-8" data-line-number="8">  <span class="dt">numericLabels =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb34-9" data-line-number="9">  <span class="dt">pamRespectsDendro =</span> <span class="ot">FALSE</span>,</a>
<a class="sourceLine" id="cb34-10" data-line-number="10">  <span class="dt">saveTOMs =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb34-11" data-line-number="11">  <span class="dt">saveTOMFileBase =</span> <span class="st">&quot;femaleMouseTOM&quot;</span>,</a>
<a class="sourceLine" id="cb34-12" data-line-number="12">  <span class="dt">verbose =</span> <span class="dv">3</span></a>
<a class="sourceLine" id="cb34-13" data-line-number="13">)</a></code></pre></div>
<pre><code>##  Calculating module eigengenes block-wise from all genes
##    Flagging genes and samples with too many missing values...
##     ..step 1
## Cluster size 3600 broken into 2108 1492 
## Cluster size 2108 broken into 1126 982 
## Done cluster 1126 
## Done cluster 982 
## Done cluster 2108 
## Done cluster 1492 
##  ..Working on block 1 .
##     TOM calculation: adjacency..
##     ..will use 6 parallel threads.
##      Fraction of slow calculations: 0.386353
##     ..connectivity..
##     ..matrix multiplication (system BLAS)..
##     ..normalization..
##     ..done.
##    ..saving TOM for block 1 into file femaleMouseTOM-block.1.RData
##  ....clustering..
##  ....detecting modules..
##  ....calculating module eigengenes..
##  ....checking kME in modules..
##      ..removing 1 genes from module 1 because their KME is too low.
##      ..removing 1 genes from module 7 because their KME is too low.
##      ..removing 1 genes from module 8 because their KME is too low.
##      ..removing 1 genes from module 21 because their KME is too low.
##  ..merging modules that are too close..
##      mergeCloseModules: Merging modules whose distance is less than 0.25
##        Calculating new MEs...</code></pre>
<p>在这里我们选择了soft thresholding power 6, 然后最小module size设置为30.<code>mergeCutHeight</code>用来控制merge module.</p>
<p>下面看一下分类结果.</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb36-1" data-line-number="1"><span class="kw">table</span>(net<span class="op">$</span>colors)</a></code></pre></div>
<pre><code>## 
##   0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18 
##  99 609 460 409 316 312 221 211 157 123 106 100  94  91  77  76  58  47  34</code></pre>
<p>代表这里面有18个module.标记为0的基因代表没有归为任何一类module.</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb38-1" data-line-number="1"><span class="co"># Convert labels to colors for plotting</span></a>
<a class="sourceLine" id="cb38-2" data-line-number="2">mergedColors =<span class="st"> </span><span class="kw">labels2colors</span>(net<span class="op">$</span>colors)</a>
<a class="sourceLine" id="cb38-3" data-line-number="3"><span class="co"># Plot the dendrogram and the module colors underneath</span></a>
<a class="sourceLine" id="cb38-4" data-line-number="4"><span class="kw">plotDendroAndColors</span>(</a>
<a class="sourceLine" id="cb38-5" data-line-number="5">  net<span class="op">$</span>dendrograms[[<span class="dv">1</span>]],</a>
<a class="sourceLine" id="cb38-6" data-line-number="6">  mergedColors[net<span class="op">$</span>blockGenes[[<span class="dv">1</span>]]],</a>
<a class="sourceLine" id="cb38-7" data-line-number="7">  <span class="st">&quot;Module colors&quot;</span>,</a>
<a class="sourceLine" id="cb38-8" data-line-number="8">  <span class="dt">dendroLabels =</span> <span class="ot">FALSE</span>,</a>
<a class="sourceLine" id="cb38-9" data-line-number="9">  <span class="dt">hang =</span> <span class="fl">0.03</span>,</a>
<a class="sourceLine" id="cb38-10" data-line-number="10">  <span class="dt">addGuide =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb38-11" data-line-number="11">  <span class="dt">guideHang =</span> <span class="fl">0.05</span></a>
<a class="sourceLine" id="cb38-12" data-line-number="12">)</a></code></pre></div>
<p><img src="r_cookbook_files/figure-html/unnamed-chunk-18-1.png" width="100%" /></p>
<p>然后将结果保存,用于后续的分析.</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb39-1" data-line-number="1">moduleLabels =<span class="st"> </span>net<span class="op">$</span>colors</a>
<a class="sourceLine" id="cb39-2" data-line-number="2">moduleColors =<span class="st"> </span><span class="kw">labels2colors</span>(net<span class="op">$</span>colors)</a>
<a class="sourceLine" id="cb39-3" data-line-number="3">MEs =<span class="st"> </span>net<span class="op">$</span>MEs</a>
<a class="sourceLine" id="cb39-4" data-line-number="4"></a>
<a class="sourceLine" id="cb39-5" data-line-number="5">geneTree =<span class="st"> </span>net<span class="op">$</span>dendrograms[[<span class="dv">1</span>]]</a>
<a class="sourceLine" id="cb39-6" data-line-number="6"></a>
<a class="sourceLine" id="cb39-7" data-line-number="7"><span class="kw">save</span>(MEs, moduleLabels, moduleColors, geneTree, </a>
<a class="sourceLine" id="cb39-8" data-line-number="8">     <span class="dt">file =</span> <span class="st">&quot;/Users/shenxt/study/WGCNA_test/FemaleLiver-02-networkConstruction-auto.RData&quot;</span>)</a></code></pre></div>
</div>
<div id="relating-modules-to-external-information-and-identifying-important" class="section level3">
<h3><span class="header-section-number">7.1.3</span> Relating modules to external information and identifying important</h3>
<p>首先加载上一步产生的数据:</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb40-1" data-line-number="1">lnames =<span class="st"> </span><span class="kw">load</span>(<span class="dt">file =</span> <span class="st">&quot;/Users/shenxt/study/WGCNA_test/FemaleLiver-01-dataInput.RData&quot;</span>);</a>
<a class="sourceLine" id="cb40-2" data-line-number="2"><span class="co">#The variable lnames contains the names of loaded variables.</span></a>
<a class="sourceLine" id="cb40-3" data-line-number="3">lnames</a></code></pre></div>
<pre><code>## [1] &quot;datExpr&quot;   &quot;datTraits&quot;</code></pre>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb42-1" data-line-number="1"><span class="co"># Load network data saved in the second part.</span></a>
<a class="sourceLine" id="cb42-2" data-line-number="2">lnames =<span class="st"> </span><span class="kw">load</span>(<span class="dt">file =</span> <span class="st">&quot;/Users/shenxt/study/WGCNA_test/FemaleLiver-02-networkConstruction-auto.RData&quot;</span>);</a>
<a class="sourceLine" id="cb42-3" data-line-number="3">lnames</a></code></pre></div>
<pre><code>## [1] &quot;MEs&quot;          &quot;moduleLabels&quot; &quot;moduleColors&quot; &quot;geneTree&quot;</code></pre>
<p>然后我们想要找到和clinical data关系非常紧密的module.对于每个module的来说,都有一个eigengene,也就是其PCA分析的第一主成分,用来代表该module在每个样品中的含量.</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb44-1" data-line-number="1"><span class="co"># Define numbers of genes and samples</span></a>
<a class="sourceLine" id="cb44-2" data-line-number="2">nGenes =<span class="st"> </span><span class="kw">ncol</span>(datExpr)</a>
<a class="sourceLine" id="cb44-3" data-line-number="3">nSamples =<span class="st"> </span><span class="kw">nrow</span>(datExpr)</a>
<a class="sourceLine" id="cb44-4" data-line-number="4"><span class="co"># Recalculate MEs with color labels</span></a>
<a class="sourceLine" id="cb44-5" data-line-number="5">MEs0 =<span class="st"> </span><span class="kw">moduleEigengenes</span>(datExpr, moduleColors)<span class="op">$</span>eigengenes</a>
<a class="sourceLine" id="cb44-6" data-line-number="6">MEs =<span class="st"> </span><span class="kw">orderMEs</span>(MEs0)</a>
<a class="sourceLine" id="cb44-7" data-line-number="7">moduleTraitCor =<span class="st"> </span><span class="kw">cor</span>(MEs, datTraits, <span class="dt">use =</span> <span class="st">&quot;p&quot;</span>)</a>
<a class="sourceLine" id="cb44-8" data-line-number="8">moduleTraitPvalue =<span class="st"> </span><span class="kw">corPvalueStudent</span>(moduleTraitCor, nSamples)</a></code></pre></div>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb45-1" data-line-number="1"><span class="co"># Will display correlations and their p-values</span></a>
<a class="sourceLine" id="cb45-2" data-line-number="2">textMatrix =<span class="st"> </span><span class="kw">paste</span>(<span class="kw">signif</span>(moduleTraitCor, <span class="dv">2</span>), <span class="st">&quot;</span><span class="ch">\n</span><span class="st">(&quot;</span>,</a>
<a class="sourceLine" id="cb45-3" data-line-number="3"><span class="kw">signif</span>(moduleTraitPvalue, <span class="dv">1</span>), <span class="st">&quot;)&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>);</a>
<a class="sourceLine" id="cb45-4" data-line-number="4"><span class="kw">dim</span>(textMatrix) =<span class="st"> </span><span class="kw">dim</span>(moduleTraitCor)</a>
<a class="sourceLine" id="cb45-5" data-line-number="5"><span class="kw">par</span>(<span class="dt">mar =</span> <span class="kw">c</span>(<span class="dv">6</span>, <span class="fl">8.5</span>, <span class="dv">3</span>, <span class="dv">3</span>));</a>
<a class="sourceLine" id="cb45-6" data-line-number="6"><span class="co"># Display the correlation values within a heatmap plot</span></a>
<a class="sourceLine" id="cb45-7" data-line-number="7"><span class="kw">labeledHeatmap</span>(<span class="dt">Matrix =</span> moduleTraitCor,</a>
<a class="sourceLine" id="cb45-8" data-line-number="8"><span class="dt">xLabels =</span> <span class="kw">names</span>(datTraits),</a>
<a class="sourceLine" id="cb45-9" data-line-number="9"><span class="dt">yLabels =</span> <span class="kw">names</span>(MEs),</a>
<a class="sourceLine" id="cb45-10" data-line-number="10"><span class="dt">ySymbols =</span> <span class="kw">names</span>(MEs),</a>
<a class="sourceLine" id="cb45-11" data-line-number="11"><span class="dt">colorLabels =</span> <span class="ot">FALSE</span>,</a>
<a class="sourceLine" id="cb45-12" data-line-number="12"><span class="dt">colors =</span> <span class="kw">greenWhiteRed</span>(<span class="dv">50</span>),</a>
<a class="sourceLine" id="cb45-13" data-line-number="13"><span class="dt">textMatrix =</span> textMatrix,</a>
<a class="sourceLine" id="cb45-14" data-line-number="14"><span class="dt">setStdMargins =</span> <span class="ot">FALSE</span>,</a>
<a class="sourceLine" id="cb45-15" data-line-number="15"><span class="dt">cex.text =</span> <span class="fl">0.5</span>,</a>
<a class="sourceLine" id="cb45-16" data-line-number="16"><span class="dt">zlim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">1</span>,<span class="dv">1</span>),</a>
<a class="sourceLine" id="cb45-17" data-line-number="17"><span class="dt">main =</span> <span class="kw">paste</span>(<span class="st">&quot;Module-trait relationships&quot;</span>))</a></code></pre></div>
<pre><code>## Warning in greenWhiteRed(50): WGCNA::greenWhiteRed: this palette is not suitable for people
## with green-red color blindness (the most common kind of color blindness).
## Consider using the function blueWhiteRed instead.</code></pre>
<p><img src="r_cookbook_files/figure-html/unnamed-chunk-22-1.png" width="100%" /></p>
<p>Gene relationship to trait and important modules: Gene Significance and Module Membership</p>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb47-1" data-line-number="1"><span class="co"># Define variable weight containing the weight column of datTrait</span></a>
<a class="sourceLine" id="cb47-2" data-line-number="2">weight =<span class="st"> </span><span class="kw">as.data.frame</span>(datTraits<span class="op">$</span>weight_g)</a>
<a class="sourceLine" id="cb47-3" data-line-number="3"><span class="kw">names</span>(weight) =<span class="st"> &quot;weight&quot;</span></a>
<a class="sourceLine" id="cb47-4" data-line-number="4"><span class="co"># names (colors) of the modules</span></a>
<a class="sourceLine" id="cb47-5" data-line-number="5">modNames =<span class="st"> </span><span class="kw">substring</span>(<span class="kw">names</span>(MEs), <span class="dv">3</span>)</a>
<a class="sourceLine" id="cb47-6" data-line-number="6">geneModuleMembership =<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">cor</span>(datExpr, MEs, <span class="dt">use =</span> <span class="st">&quot;p&quot;</span>))</a>
<a class="sourceLine" id="cb47-7" data-line-number="7">MMPvalue =<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">corPvalueStudent</span>(<span class="kw">as.matrix</span>(geneModuleMembership), nSamples))</a>
<a class="sourceLine" id="cb47-8" data-line-number="8"><span class="kw">names</span>(geneModuleMembership) =<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;MM&quot;</span>, modNames, <span class="dt">sep=</span><span class="st">&quot;&quot;</span>)</a>
<a class="sourceLine" id="cb47-9" data-line-number="9"><span class="kw">names</span>(MMPvalue) =<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;p.MM&quot;</span>, modNames, <span class="dt">sep=</span><span class="st">&quot;&quot;</span>)</a>
<a class="sourceLine" id="cb47-10" data-line-number="10">geneTraitSignificance =<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">cor</span>(datExpr, weight, <span class="dt">use =</span> <span class="st">&quot;p&quot;</span>))</a>
<a class="sourceLine" id="cb47-11" data-line-number="11">GSPvalue =<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">corPvalueStudent</span>(<span class="kw">as.matrix</span>(geneTraitSignificance), nSamples))</a>
<a class="sourceLine" id="cb47-12" data-line-number="12"><span class="kw">names</span>(geneTraitSignificance) =<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;GS.&quot;</span>, <span class="kw">names</span>(weight), <span class="dt">sep=</span><span class="st">&quot;&quot;</span>)</a>
<a class="sourceLine" id="cb47-13" data-line-number="13"><span class="kw">names</span>(GSPvalue) =<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;p.GS.&quot;</span>, <span class="kw">names</span>(weight), <span class="dt">sep=</span><span class="st">&quot;&quot;</span>)</a></code></pre></div>
<p>Intramodular analysis: identifying genes with high GS and MM</p>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb48-1" data-line-number="1">module =<span class="st"> &quot;brown&quot;</span></a>
<a class="sourceLine" id="cb48-2" data-line-number="2">column =<span class="st"> </span><span class="kw">match</span>(module, modNames)</a>
<a class="sourceLine" id="cb48-3" data-line-number="3">moduleGenes =<span class="st"> </span>moduleColors<span class="op">==</span>module</a>
<a class="sourceLine" id="cb48-4" data-line-number="4"><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</a>
<a class="sourceLine" id="cb48-5" data-line-number="5"><span class="kw">verboseScatterplot</span>(<span class="kw">abs</span>(geneModuleMembership[moduleGenes, column]),</a>
<a class="sourceLine" id="cb48-6" data-line-number="6">                   <span class="kw">abs</span>(geneTraitSignificance[moduleGenes, <span class="dv">1</span>]),</a>
<a class="sourceLine" id="cb48-7" data-line-number="7"><span class="dt">xlab =</span> <span class="kw">paste</span>(<span class="st">&quot;Module Membership in&quot;</span>, module, <span class="st">&quot;module&quot;</span>),</a>
<a class="sourceLine" id="cb48-8" data-line-number="8"><span class="dt">ylab =</span> <span class="st">&quot;Gene significance for body weight&quot;</span>,</a>
<a class="sourceLine" id="cb48-9" data-line-number="9"><span class="dt">main =</span> <span class="kw">paste</span>(<span class="st">&quot;Module membership vs. gene significance</span><span class="ch">\n</span><span class="st">&quot;</span>),</a>
<a class="sourceLine" id="cb48-10" data-line-number="10"><span class="dt">cex.main =</span> <span class="fl">1.2</span>, <span class="dt">cex.lab =</span> <span class="fl">1.2</span>, <span class="dt">cex.axis =</span> <span class="fl">1.2</span>, <span class="dt">col =</span> module)</a></code></pre></div>
<p><img src="r_cookbook_files/figure-html/unnamed-chunk-24-1.png" width="100%" /></p>
<p>Summary output of network analysis results</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb49-1" data-line-number="1"><span class="kw">head</span>(<span class="kw">names</span>(datExpr))</a></code></pre></div>
<pre><code>## [1] &quot;MMT00000044&quot; &quot;MMT00000046&quot; &quot;MMT00000051&quot; &quot;MMT00000076&quot; &quot;MMT00000080&quot; &quot;MMT00000102&quot;</code></pre>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb51-1" data-line-number="1"><span class="kw">head</span>(<span class="kw">names</span>(datExpr)[moduleColors<span class="op">==</span><span class="st">&quot;brown&quot;</span>])</a></code></pre></div>
<pre><code>## [1] &quot;MMT00000887&quot; &quot;MMT00001077&quot; &quot;MMT00001185&quot; &quot;MMT00001486&quot; &quot;MMT00002002&quot; &quot;MMT00002037&quot;</code></pre>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb53-1" data-line-number="1">annot &lt;-<span class="st"> </span></a>
<a class="sourceLine" id="cb53-2" data-line-number="2"><span class="st">  </span><span class="kw">read.csv</span>(<span class="dt">file =</span> <span class="st">&quot;/Users/shenxt/study/WGCNA_test/GeneAnnotation.csv&quot;</span>)</a>
<a class="sourceLine" id="cb53-3" data-line-number="3"><span class="kw">dim</span>(annot)</a></code></pre></div>
<pre><code>## [1] 23388    34</code></pre>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb55-1" data-line-number="1"><span class="kw">names</span>(annot)</a></code></pre></div>
<pre><code>##  [1] &quot;X&quot;                        &quot;ID&quot;                       &quot;arrayname&quot;                &quot;substanceBXH&quot;            
##  [5] &quot;gene_symbol&quot;              &quot;LocusLinkID&quot;              &quot;OfficialGeneSymbol&quot;       &quot;OfficialGeneName&quot;        
##  [9] &quot;LocusLinkSymbol&quot;          &quot;LocusLinkName&quot;            &quot;ProteomeShortDescription&quot; &quot;UnigeneCluster&quot;          
## [13] &quot;LocusLinkCode&quot;            &quot;ProteomeID&quot;               &quot;ProteomeCode&quot;             &quot;SwissprotID&quot;             
## [17] &quot;OMIMCode&quot;                 &quot;DirectedTilingPriority&quot;   &quot;AlternateSymbols&quot;         &quot;AlternateNames&quot;          
## [21] &quot;SpeciesID&quot;                &quot;cytogeneticLoc&quot;           &quot;Organism&quot;                 &quot;clustername&quot;             
## [25] &quot;reporterid&quot;               &quot;probeid&quot;                  &quot;sequenceid&quot;               &quot;clusterid&quot;               
## [29] &quot;chromosome&quot;               &quot;startcoordinate&quot;          &quot;endcoordinate&quot;            &quot;strand&quot;                  
## [33] &quot;sequence_3_to_5_prime&quot;    &quot;sequence_5_to_3_prime&quot;</code></pre>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb57-1" data-line-number="1">probes =<span class="st"> </span><span class="kw">names</span>(datExpr)</a>
<a class="sourceLine" id="cb57-2" data-line-number="2">probes2annot =<span class="st"> </span><span class="kw">match</span>(probes, annot<span class="op">$</span>substanceBXH)</a>
<a class="sourceLine" id="cb57-3" data-line-number="3"><span class="co"># The following is the number or probes without annotation:</span></a>
<a class="sourceLine" id="cb57-4" data-line-number="4"><span class="kw">sum</span>(<span class="kw">is.na</span>(probes2annot))</a></code></pre></div>
<pre><code>## [1] 0</code></pre>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb59-1" data-line-number="1"><span class="co"># Create the starting data frame</span></a>
<a class="sourceLine" id="cb59-2" data-line-number="2">geneInfo0 =<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">substanceBXH =</span> probes,</a>
<a class="sourceLine" id="cb59-3" data-line-number="3"><span class="dt">geneSymbol =</span> annot<span class="op">$</span>gene_symbol[probes2annot],</a>
<a class="sourceLine" id="cb59-4" data-line-number="4"><span class="dt">LocusLinkID =</span> annot<span class="op">$</span>LocusLinkID[probes2annot],</a>
<a class="sourceLine" id="cb59-5" data-line-number="5"><span class="dt">moduleColor =</span> moduleColors,</a>
<a class="sourceLine" id="cb59-6" data-line-number="6">geneTraitSignificance,</a>
<a class="sourceLine" id="cb59-7" data-line-number="7">GSPvalue)</a>
<a class="sourceLine" id="cb59-8" data-line-number="8"><span class="co"># Order modules by their significance for weight</span></a>
<a class="sourceLine" id="cb59-9" data-line-number="9">modOrder =<span class="st"> </span><span class="kw">order</span>(<span class="op">-</span><span class="kw">abs</span>(<span class="kw">cor</span>(MEs, weight, <span class="dt">use =</span> <span class="st">&quot;p&quot;</span>)));</a>
<a class="sourceLine" id="cb59-10" data-line-number="10"><span class="co"># Add module membership information in the chosen order</span></a>
<a class="sourceLine" id="cb59-11" data-line-number="11"><span class="cf">for</span> (mod <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(geneModuleMembership))</a>
<a class="sourceLine" id="cb59-12" data-line-number="12">{</a>
<a class="sourceLine" id="cb59-13" data-line-number="13">oldNames =<span class="st"> </span><span class="kw">names</span>(geneInfo0)</a>
<a class="sourceLine" id="cb59-14" data-line-number="14">geneInfo0 =<span class="st"> </span><span class="kw">data.frame</span>(geneInfo0, geneModuleMembership[, modOrder[mod]],</a>
<a class="sourceLine" id="cb59-15" data-line-number="15">MMPvalue[, modOrder[mod]]);</a>
<a class="sourceLine" id="cb59-16" data-line-number="16"><span class="kw">names</span>(geneInfo0) =<span class="st"> </span><span class="kw">c</span>(oldNames, <span class="kw">paste</span>(<span class="st">&quot;MM.&quot;</span>, modNames[modOrder[mod]], <span class="dt">sep=</span><span class="st">&quot;&quot;</span>),</a>
<a class="sourceLine" id="cb59-17" data-line-number="17"><span class="kw">paste</span>(<span class="st">&quot;p.MM.&quot;</span>, modNames[modOrder[mod]], <span class="dt">sep=</span><span class="st">&quot;&quot;</span>))</a>
<a class="sourceLine" id="cb59-18" data-line-number="18">}</a>
<a class="sourceLine" id="cb59-19" data-line-number="19"><span class="co"># Order the genes in the geneInfo variable first by module color, then by geneTraitSignificance</span></a>
<a class="sourceLine" id="cb59-20" data-line-number="20">geneOrder =<span class="st"> </span><span class="kw">order</span>(geneInfo0<span class="op">$</span>moduleColor, <span class="op">-</span><span class="kw">abs</span>(geneInfo0<span class="op">$</span>GS.weight));</a>
<a class="sourceLine" id="cb59-21" data-line-number="21">geneInfo =<span class="st"> </span>geneInfo0[geneOrder, ]</a>
<a class="sourceLine" id="cb59-22" data-line-number="22"><span class="kw">write.csv</span>(geneInfo, <span class="dt">file =</span> <span class="st">&quot;/Users/shenxt/study/WGCNA_test/geneInfo.csv&quot;</span>)</a></code></pre></div>
</div>
<div id="对module进行分析" class="section level3">
<h3><span class="header-section-number">7.1.4</span> 对module进行分析</h3>
<p>在前文中,鉴定出了几个和weight相关的module.</p>
<hr />
<div id="disqus_thread">

</div>
<script>

/**
*  RECOMMENDED CONFIGURATION VARIABLES: EDIT AND UNCOMMENT THE SECTION BELOW TO INSERT DYNAMIC VALUES FROM YOUR PLATFORM OR CMS.
*  LEARN WHY DEFINING THESE VARIABLES IS IMPORTANT: https://disqus.com/admin/universalcode/#configuration-variables*/
/*
var disqus_config = function () {
this.page.url = PAGE_URL;  // Replace PAGE_URL with your page's canonical URL variable
this.page.identifier = PAGE_IDENTIFIER; // Replace PAGE_IDENTIFIER with your page's unique identifier variable
};
*/
(function() { // DON'T EDIT BELOW THIS LINE
var d = document, s = d.createElement('script');
s.src = 'https://r-cookbook-shen.disqus.com/embed.js';
s.setAttribute('data-timestamp', +new Date());
(d.head || d.body).appendChild(s);
})();
</script>
<noscript>
Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a>
</noscript>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="network.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="regularexpression.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["r_cookbook.pdf", "r_cookbook.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
